src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```diff
@@ -795,22 +799,29 @@ public Task<List<RawFunctionMetadata>> GetFunctionMetadata()
 
         internal Task<List<RawFunctionMetadata>> SendFunctionMetadataRequest()
         {
-            // reset indexing task when in case we need to send another request
-            _functionsIndexingTask = new TaskCompletionSource<List<RawFunctionMetadata>>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _workerChannelLogger.LogDebug("Fetching worker metadata, FunctionMetadataReceived set to: {functionMetadataReceived}", _functionMetadataRequestSent);
+            if (!_functionMetadataRequestSent)
+            {
+                lock (_metadataLock)
+                {
```

Comment:
You need to re-check `if (!_functionMetadataRequestSent)` here inside the lock as well. It's possible that there's a race between here and the last check. So you typically (always?) need to check once outside the lock as a potential short-circuit, then once again inside where you know you're the only thread.

Reply:
Have added the addtional boolean check